File: Algorithm/Python/086. Partition List.py
# -*- coding: utf-8 -*-
# @Author: LC
# @Date:   2016-05-26 17:53:02
# @Last modified by:   WuLC
# @Last Modified time: 2016-05-26 17:53:15
# @Email: liangchaowu5@gmail.com

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        p1, ip = dummy, None
        start, end = None, None
        while ip == None and p1.next!=None:
            if p1.next.val >= x :
                ip = p1
            p1 = p1.next
            
        while p1.next != None:
            if p1.next.val < x:
                if start == None:
                    start = end = p1.next
                else:
                    end.next = p1.next
                    end = end.next
                p1.next = p1.next.next
            else:
                p1 = p1.next
            
        if ip!= None and start != None:
            end.next = ip.next
            ip.next = start
        return dummy.next

File: Algorithm/Java/086. Partition List.java
/**
* Author: WuLC
* Date:   2017-03-12 23:58:09
* Last modified by:   WuLC
* Last Modified time: 2017-03-12 23:58:56
* Email: liangchaowu5@gmail.com
*/


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */


// O(n) space, store smaller values and largers values in two seperate list
public class Solution 
{
    public ListNode partition(ListNode head, int x) 
    {
        List<Integer> smaller = new ArrayList<Integer>();
        List<Integer> larger = new ArrayList<Integer>();
        ListNode curr = head;
        while (curr != null)
        {
            if (curr.val < x) smaller.add(curr.val);
            else larger.add(curr.val);
            curr = curr.next;
        }
        
        curr = head;
        for(int tmp : smaller)
        {
            curr.val = tmp;
            curr = curr.next;
        }
        for(int tmp : larger)
        {
            curr.val = tmp;
            curr = curr.next;
        }
        return head;
    }
}